### ISSUE TITLE
[TRACKER] Unit tests to add or improve

### ISSUE THREAD

**Issue by Calinou** [godotengine/godot#43440]
Our unit test coverage is currently fairly low. We'd like to increase our unit test coverage; any help is welcome.

**Interested in writing new unit tests?** See the [unit tests documentation](https://docs.godotengine.org/en/latest/development/cpp/unit_testing.html) and [compiling instructions](https://docs.godotengine.org/en/latest/contributing/development/compiling/).
If you have further questions, join the [Godot Contributors Chat](https://chat.godotengine.org/).

When opening a pull request, please link back to this issue (`#43440`) in the PR description so that we can keep track of it more easily.

> If you have the appropriate permissions, feel free to edit this issue to add new items or check items for which a PR has been opened.

### Classes to test

*These classes are currently lacking in test coverage, and are therefore highest-priority for receiving unit tests. Deprecated classes are not listed.*

> [!NOTE]
>
> When a class is listed with "and" along a given list item, it should be submitted in the same pull request whenever possible. Tests for these classes can be in the same file or a different file depending on the size and complexity of the test suite. If in doubt, follow the file organization used in the original class implementation.

- [x] **AnimationPlayer** https://github.com/godotengine/godot/pull/92649
- [ ] **AudioStreamInteractive** (and **AudioStreamPlaybackInteractive**, **AudioStreamPlaylist**, **AudioStreamSynchronized**)
- [x] **BaseButton** (and **Button**) https://github.com/godotengine/godot/pull/93381
- [ ] **Control:** Test ~~sizing, anchors and~~ container logic by checking control position and size after creating an UI layout. Test focus neighbor behavior using simulated input events (`ui_focus_` actions). This class is complex, so tests for it can be split in multiple pull requests.
  - https://github.com/godotengine/godot/pull/98757
- [x] **Decal** https://github.com/godotengine/godot/pull/93463
- [ ] **ENetConnection** (and **ENetPacketPeer**)
- [x] **FontFile** https://github.com/godotengine/godot/pull/99131
- [x] **FontVariation** https://github.com/godotengine/godot/pull/99687
- [x] **GLTFDocument** https://github.com/godotengine/godot/pull/101136
- [x] **HeightMapShape3D** https://github.com/godotengine/godot/pull/95123
- [x] **HTTPRequest** https://github.com/godotengine/godot/pull/95224
- [x] **ImageTexture3D** https://github.com/godotengine/godot/pull/91233
- [x] **Node2D** https://github.com/godotengine/godot/pull/91654
- [x] **Node3D** https://github.com/godotengine/godot/pull/97143
- [x] **Occluder3D** (and **ArrayOccluder3D**, **BoxOccluder3D**, **PolygonOccluder3D**, **QuadOccluder3D**, and **SphereOccluder3D**) https://github.com/godotengine/godot/pull/99437
- [ ] **OccluderInstance3D:** *Currently being worked on by @carsonetb*
- [ ] **OptimizedTranslation** *Currently being worked on by @SommAid*
- [x] **OptionButton** https://github.com/godotengine/godot/pull/93824
- [x] **PacketPeer** https://github.com/godotengine/godot/pull/95931
- [x] **Parallax2D** https://github.com/godotengine/godot/pull/95650
- [x] **PhysicsMaterial** https://github.com/godotengine/godot/pull/97227
- [x] **ReflectionProbe** https://github.com/godotengine/godot/pull/97818 https://github.com/godotengine/godot/pull/97870
- [x] **Sky** https://github.com/godotengine/godot/pull/99886 https://github.com/godotengine/godot/pull/97512 
- [ ] **StandardMaterial3D** (and **ORMMaterial3D**)
- [ ] **StreamPeerTCP** *Currently being worked on by @wheatear-dev*
- [ ] **StreamPeerTLS**
- [ ] **StreamPeerGZIP** *Currently being worked on by @tsarquis88*
- [x] **StreamPeerBuffer** https://github.com/godotengine/godot/pull/100871 https://github.com/godotengine/godot/pull/95784
- [ ] **StyleBoxFlat** *Currently being worked on by @mshojatalab*
- [ ] **StyleBoxLine** (and **StyleBoxEmpty**) *Currently being worked on by @rcooper47*
- [x] **StyleBoxTexture** https://github.com/godotengine/godot/pull/95218
- [x] **Timer** https://github.com/godotengine/godot/pull/97794
- [ ] **WebSocketPeer** *Currently being worked on by @CampioniMan*
- [ ] **VisibleOnScreenNotifier2D:** May not be testable in headless mode, but it's worth trying.
- [ ] **VisibleOnScreenNotifier3D:** May not be testable in headless mode, but it's worth trying.

### Completed classes

*These classes currently have good test coverage. Further improvements may be possible by testing methods that were added after the tests were merged.*

<details>

- [x] **AABB** https://github.com/godotengine/godot/pull/43727
- [x] **Animation** https://github.com/godotengine/godot/pull/57490
- [x] **Array** https://github.com/godotengine/godot/pull/46227
- [x] **ArrayMesh**
- [x] **AudioStreamWAV** https://github.com/godotengine/godot/pull/60736
- [x] **BitMap**
- [x] **Camera2D** https://github.com/godotengine/godot/pull/88614
- [x] **Camera3D:** May not be testable in headless mode, but it's worth trying. Test `is_position_behind()` and the `project_`/`unproject_` methods in particular
- [x] **CryptoKey** https://github.com/godotengine/godot/pull/89021
- [x] **Curve** https://github.com/godotengine/godot/pull/43516
- [x] **Curve2D** https://github.com/godotengine/godot/pull/71308
- [x] **Curve3D** #76812
- [x] **Dictionary** https://github.com/godotengine/godot/pull/47511
- [x] **FastNoiseLite**
- [x] **Font** https://github.com/godotengine/godot/pull/81503
- [x] **GDScript** https://github.com/godotengine/godot/pull/48657
- [x] **Geometry2D** https://github.com/godotengine/godot/pull/59643
- [x] **Geometry3D** https://github.com/godotengine/godot/pull/44974
- [x] **GradientTexture** https://github.com/godotengine/godot/pull/90501
- [x] **HashingContext** https://github.com/godotengine/godot/pull/43459
- [x] **HTTPClient** https://github.com/godotengine/godot/pull/76636
- [x] **Image** https://github.com/godotengine/godot/pull/45737
- [x] **ImageTexture** https://github.com/godotengine/godot/pull/88044
- [x] **InputEvent** https://github.com/godotengine/godot/pull/79444
- [x] **InputEventKey** https://github.com/godotengine/godot/pull/59193
- [x] **JSON** https://github.com/godotengine/godot/pull/43517
- [x] **JSONRPC**: https://github.com/godotengine/godot/pull/89124
- [x] **Math** https://github.com/godotengine/godot/pull/48721
- [x] **Marshalls** https://github.com/godotengine/godot/pull/44797
- [x] **Node**: https://github.com/godotengine/godot/pull/70152 https://github.com/godotengine/godot/pull/71367
- [x] **NodePath** https://github.com/godotengine/godot/pull/43578
- [x] **NoiseTexture** https://github.com/godotengine/godot/pull/70919
- [x] **Object** https://github.com/godotengine/godot/pull/43583 (completed by https://github.com/godotengine/godot/pull/44387)
- [x] **OS** https://github.com/godotengine/godot/pull/42069
- [x] **PackedScene** https://github.com/godotengine/godot/pull/79440 https://github.com/godotengine/godot/pull/80423 https://github.com/godotengine/godot/pull/80819
- [x] **Path2D** https://github.com/godotengine/godot/pull/66927
- [x] **PathFollow2D** https://github.com/godotengine/godot/pull/46277
- [x] **Path3D**
- [x] **PathFollow3D** https://github.com/godotengine/godot/pull/46341
- [x] **PCKPacker** https://github.com/godotengine/godot/pull/43572
- [x] **Plane** https://github.com/godotengine/godot/pull/44492
- [x] **PrimitiveMesh** (CapsuleMesh, CubeMesh, …)
- [x] **Projection** https://github.com/godotengine/godot/pull/77275 https://github.com/godotengine/godot/pull/99125
- [x] **Quaternion** https://github.com/godotengine/godot/pull/50907
- [x] **RandomNumberGenerator** https://github.com/godotengine/godot/pull/44560
- [x] **Rect2:** Not covered in the legacy math test. https://github.com/godotengine/godot/pull/43514
- [x] **Rect2i:** Not covered in the legacy math test. https://github.com/godotengine/godot/pull/43514
- [x] **RegEx** https://github.com/godotengine/godot/pull/82225
- [x] **RID** https://github.com/godotengine/godot/pull/54325
- [x] **Resource** https://github.com/godotengine/godot/pull/43731
- [x] **Shortcut** https://github.com/godotengine/godot/pull/58712
- [x] **SpriteFrames** https://github.com/godotengine/godot/pull/57742
- [x] **Transform**
- [x] **Transform2D**
- [x] **Translation** https://github.com/godotengine/godot/pull/48778
- [x] **TranslationServer** https://github.com/godotengine/godot/pull/79331
- [x] **Variant** https://github.com/godotengine/godot/pull/76244
- [x] **Vector** https://github.com/godotengine/godot/pull/48821
- [x] **Vector2/Vector2i** https://github.com/godotengine/godot/pull/47202
- [x] **Vector3/Vector3i** https://github.com/godotengine/godot/pull/48063
- [x] **Vector4/Vector4i** https://github.com/godotengine/godot/pull/64027
- [x] **VisualShader** https://github.com/godotengine/godot/pull/70396
</details>

### Non-testable classes

*These classes can't be unit-tested for technical reasons. Unit tests always run in headless mode, so they can't do things such as rendering scenes and checking the visual result.*

<details>

- [x] ~~**CubeMap:**~~ Not testable without RenderingServer access.
- [x] ~~**Shader:**~~ Not testable without RenderingServer access.
</details>

**Comment by singher** [comment]
Hi me and a friend would like to take up some of these unit tests for a class project we have. Is there someone we can stay in touch with if we have questions?

**Comment by Calinou** [comment]
@singher Thanks for your interest in contributing! I recommend joining the [`#godotengine-devel` IRC channel on Freenode](https://webchat.freenode.net/?channels=#godotengine).

**Comment by singher** [comment]
@Calinou perfect! Thank you so much for this opportunity. We will try our best to cover as many of those classes to test.

**Comment by vinayakmtiwari** [comment]
@Calinou Hi, I am working with @singher on this. We just wanted to let you know that we were planning on writing test cases for the following classes: Array, Dictionary, Node, RandomNumberGenerator.

**Comment by Xrayez** [comment]
> we were planning on writing test cases for the following classes: Array, Dictionary, Node, RandomNumberGenerator.

@vinayakmtiwari `RandomNumberGenerator` initial test suite is written in #43103 and #43343 for new functionality, so I guess you could take it as a base.

**Comment by MaxMutantMayer** [comment]
I would like to tackle some of the test cases too and would start with the classes `HashingContext` and `RegEx`.
Do you prefer separate PRs for each class, or should I aggregate them?

Ps.: First time contributor here!

**Comment by Xrayez** [comment]
@MaxMutantMayer it's better to create separate PRs for each class, this eases the review process → potentially speeds up the merging process. 😉

**Comment by MaxMutantMayer** [comment]
@Xrayez Alright, makes sense. Thanks for clarification :relaxed:

**Comment by jonbonazza** [comment]
Noting for posterity that @Calinou found several issues with the JSON facilities not conforming to the spec. He purposely left these unit tests out of the PR to not confuse people, but I created a proposal to address this: https://github.com/godotengine/godot-proposals/issues/1833

**Comment by Gorgonx7** [comment]
Hey if you don't mind I'm gonna take a crack at testing geometry 3D, it's also my first time contributing too :D

**Comment by Xrayez** [comment]
I had to write test suite for `RandomNumberGenerator` in #44089 with test cases for reproducibility, but not all public methods are currently tested.

**Comment by rmarco3e8** [comment]
Hello, I'm also looking to start contributing and am writing tests for `Path2D`. If someone has already been working on that class, let me know. Thanks for the resources dropped earlier. I'll be sure to ask any questions in the Freenode channel.

**Comment by cabinboy1031** [comment]
I am going to see about writing unit tests for the `Plane` class. Though this is my first contribution to anything in Github. If someone more experienced than me ends up wanting to do the unit tests go ahead as mine will probably not be the best and will most likely need to be improved after i am done with it anyways.

**Comment by vinayakmtiwari** [comment]
Hey @Calinou, I added some unit test cases for the Random Number Generator class in [#44350 ]. Also, I would like to remove my name and @singher from contributors to the Node class

**Comment by rmarco3e8** [comment]
Hey, I opened another PR for `Object` with a few more test cases in #44387.

**Comment by a-ivanov** [comment]
Hi there! I want to make my first time contribution to the project by writing unit tests for `marshalls.h` functions.

**Comment by Calinou** [comment]
@a-ivanov Nobody started work on Marshalls yet, but I haven't been able to access that class from C++ since it's mostly intended to be used from GDScript.

**Comment by Xrayez** [comment]
I think it should be possible to test in either case, because core singletons are initialized in the test environment, I don't know exactly what might be a limitation for doing so.

For `_Marshalls` in `core_bind.h` you may just need to use `_Marshalls::get_singleton()->variant_to_base64()...` etc, but again not sure whether bind classes have to be tested currently, I think it may be better to test core stuff, which may be even easier to do if core doesn't rely on `ClassDB` functionality (those classes which expand `GDCLASS` macro), and just rely on pure data manipulation, which should be the case for marshalls I presume.

**Comment by a-ivanov** [comment]
> Nobody has started work on testing Marshalls yet, but I haven't been able to access that class from C++ since it's mostly intended to be used from GDScript. It might not be easy to test.

Actually I thought about the strategy suggested by @Xrayez:

1. start with testing global functions in `marshalls.h`: `encode_uint16`, `encode_uint32`, etc. - first PR,
2. test `_Marshalls` class (need additional research on how to approach it, start with what @Xrayez wrote) - another PR.

Anyway, gonna ask you details later here/in Freenode channel.

Thanks.

**Comment by BrooklynWelsh** [comment]
Hello everyone, interested in making my first contribution with tests for the `Node` class. I've written a couple simple tests so far, have a question or two for the Freenode channel, but going well so far. Just wanted to comment to make sure it's not being worked on.

**Comment by Xrayez** [comment]
Since we talk about `Node` classes, I think it's also possible to simulate `_process()` calls with
`node->notification(NOTIFICATION_PROCESS)`, or even `propagate_notification()`, for those classes which are supposed to be added to the scene tree (most of them)... Process notification are done in a main loop I presume, and current test environment has no `SceneTree` instantiated. It may also be possible to instantiate `SceneTree` manually (and this will likely reveal some hidden bugs).

Those are just some considerations if someone hadn't have much luck with testing `Node` derived classes yet...

**Comment by BrooklynWelsh** [comment]
@Xrayez  Thanks for mentioning that! I've been meaning to ask more questions in the IRC chat but have been busy the past couple days.

I have indeed been having problems testing the functions which require `SceneTree`.

You mention instantiating the `SceneTree` manually, could you tell me what the correct way is of doing that? During my search on how to add a `Node` to a `SceneTree`, I tried a few methods of instantiating one manually but all gave me errors, with the stack trace leading to an error with `RenderingServer` in a file somewhere (it's been a couple days and I can't remember where).

**Comment by Xrayez** [comment]
Rendering is disabled in the test environment (`RenderingServer` is not instantiated at all currently) so likely `SceneTree` depends on rendering functionality in some way (`Viewport`?), leading to crash. If you look at changes in #40980, you'll see that I've added some checks in different parts of the engine that were dependent on rendering code, so something similar needs to be done to fix this in the `SceneTree` or other relevant classes. 

This kind of a fix would only work for really simple limitations, otherwise we may need to instantiate `RenderingServer` functionality eventually, yet the idea behind unit testing is that we should only use what's needed... See also godotengine/godot-proposals#1760.

That said, unless you'd like to fix those issues yourself, the only option for now is to use `node->propagate_notification(NOTIFICATION_PROCESS)` and alike...

**Comment by a-ivanov** [comment]
Hey, everyone! Gonna add a few more tests for `Object` class.

**Comment by Hyperion101010** [comment]
@Calinou  I would like to work on Curve2D test, 
I am a new contributor with experience writting C++ code.

**Comment by Calinou** [comment]
@Hyperion101010 Sure, go ahead :slightly_smiling_face:

**Comment by a-ivanov** [comment]
Do you think it makes sense to try to write unit tests for Mono-related classes like `CSharpLanguage`, `CSharpScript`, etc.?

**Comment by Calinou** [comment]
> Do you think it makes sense to try to write unit tests for Mono-related classes like CSharpLanguage, CSharpScript, etc.?

Sure, as long as these tests work when Mono is disabled or aren't compiled when Mono is disabled.

**Comment by Xrayez** [comment]
You can look at GDNative tests for reference (which is a module): https://github.com/godotengine/godot/tree/master/modules/gdnative/tests

> Sure, as long as these tests work when Mono is disabled (or aren't compiled when Mono is disabled).

I think this shouldn't be a problem since it's for development purposes, but I don't recall whether those module-specific tests are compiled or not when disabled.

**Comment by OverloadedOrama** [comment]
Hello. I'd like to try writing tests for the `Image` class. I've already cloned the repo and started writing some tests for image creation, file saving and loading and drawing pixels (with `set_pixel()`) and everything seems to be working fine so far. My main issue is that I am very new to C++ and writing unit tests in general, so I might make mistakes, and I'm not entirely sure what kind of tests each class needs. Should I add as many as I possibly can think of and then open a PR?

**Comment by Xrayez** [comment]
@OverloadedOrama you should strive good [coverage](https://en.wikipedia.org/wiki/Code_coverage) when writing unit tests. It means that you can focus on testing every **public** method available in the `Image` class. You can also go the "use case" way (find a typical, real-world usage of the class, but which would be simple enough to implement).

I'd say the Godot's source code has a lot of examples to learn from.

**Comment by Calinou** [comment]
@OverloadedOrama @Xrayez I tried to test Image but got a crash due to pthreads whenever I loaded a PNG or JPEG image from a buffer…

**Comment by OverloadedOrama** [comment]
> @OverloadedOrama @Xrayez [I tried to test Image](https://github.com/Calinou/godot/tree/test-add-image) but got a crash due to pthreads whenever I loaded a PNG or JPEG image from a buffer…

@Calinou @Xrayez I gave [loading images from buffers a try as well](https://github.com/OverloadedOrama/godot/commit/07746c58fb36f557eba9643109482b558463d082), and they seem to be running fine in my local Windows 10 machine. All successes, no crashes. The rest of the tests seem also to be running fine, and I'll try to add more before opening a PR.

**Comment by sps1112** [comment]
Hey everyone, i would like to write tests for the ``Variant`` class. The class seems to already have some tests. What other tests should it have? Should i write tests for every combination of data-types? 
This is also my first time contributing to an organisation but i have some experience writing C++ code.

**Comment by Calinou** [comment]
> What other tests should it have? Should i write tests for every combination of data-types?

As I said in the first post, I think it's important to test all the operator overloads on as many Variant types as we can.

**Comment by sps1112** [comment]
@Calinou Sure. I will try to check as many combinations as possible.

**Comment by angad-k** [comment]
I'd like to try my hand at writing some of the unit tests...

**Comment by Calinou** [comment]
> I'd like to try my hand at writing some of the unit tests... I'd like to write tests for PathFollow2D but would require some pointers on where to start from and what all to check?

I think the PathFollow2D test should:

- Create a Path2D resource and add some points to it.
- Get the position at specified points of the curve (for instance, at offsets `-0.2`, `0.0`, `0.2`, `0.6`, `1.0`, `1.2`) and make sure it matches the expected value.
  - Some of the mentioned offsets may be invalid, so make sure a proper value is returned in this case.
  - Once you're confident with your test code, you can silence the error messages by surrounding erroneous blocks of code with the `ERR_PRINT_OFF;` and `ERR_PRINT_ON;` macros.
- Remove one point from the Path2D and try getting the position at the same offsets as before. See if correct values are being returned (these are expected to be different from the previous ones).
- Remove all points from the Path2D, try to remove a point while the Path2D is empty and see what happens. The engine should print an error message but shouldn't crash :slightly_smiling_face:

**Comment by angad-k** [comment]
@Calinou thanks a lot for explaining! I'll start working on this from tomorrow and get back to you if I face any problem.

**Comment by angad-k** [comment]
@Calinou next I'd like to work on PathFollow3D. I guess the flow would overall be similar to PathFollow2D except for the addition of one more dimension, correct?

**Comment by Calinou** [comment]
@angad-k Indeed, it should be pretty similar.

**Comment by angad-k** [comment]
Okay. Will send a PR in a day or two...

**Comment by JoseTMonagas** [comment]
Hi, I want to contribute, this would be my first contribution in Godot and in open source projects in general so I may stumble around a bit, I want to work on the unit tests for Vector, I'm guessing that means Vector2 and Vector3 in core/math, right? Anything I should keep in mind?

**Comment by Calinou** [comment]
> Hi, I want to contribute, this would be my first contribution in Godot and in open source projects in general so I may stumble around a bit, I want to work on the unit tests for Vector, I'm guessing that means Vector2 and Vector3 in core/math, right? Anything I should keep in mind?

There was a small test suite for Vector2 and Vector3 in the [legacy `test_math.cpp`](https://github.com/godotengine/godot/blob/master/tests/test_math.cpp) file (this file isn't run in the current test suite). However, as you can see, it's not quite complete and it's unorganized. Therefore, I recommend starting from scratch at this point.

In OP, I also referred to the Vector *datatype* (an array-like structure), but this is not the same as Vector2 and Vector3. Nonetheless, tests for Vector, Vector2 and Vector3 are all welcome :slightly_smiling_face:

**Comment by JoseTMonagas** [comment]
> > Hi, I want to contribute, this would be my first contribution in Godot and in open source projects in general so I may stumble around a bit, I want to work on the unit tests for Vector, I'm guessing that means Vector2 and Vector3 in core/math, right? Anything I should keep in mind?
> 
> There was a small test suite for Vector2 and Vector3 in the [legacy `test_math.cpp`](https://github.com/godotengine/godot/blob/master/tests/test_math.cpp) file (this file isn't run in the current test suite). However, as you can see, it's not quite complete and it's unorganized. Therefore, I recommend starting from scratch at this point.
> 
> In OP, I also referred to the Vector _datatype_ (an array-like structure), but this is not the same as Vector2 and Vector3. Nonetheless, tests for Vector, Vector2 and Vector3 are all welcome

Great, I started working on it, at least for Vector2, and also as mentioned by @Xrayez here:

> it's better to create separate PRs for each class, this eases the review process → potentially speeds up the merging process. 

As Vector2 and Vector3 are in different files, should I make different PRs for them? The same would be for the Vector datatype which I'm assuming is the one located in core/templates/vector.h, right?

**Comment by Xrayez** [comment]
@JoseTMonagas I think the rule of thumb is when you see a source file, it can have a corresponding test suite, so yeah makes sense to create separate PRs for Vector2/Vector3.

**Comment by resul4e** [comment]
@Calinou Could you maybe check Geometry2D because I thought there were no tests yet for this class, but it seems that @mbrlabs has already added tests in #44274.

**Comment by mbrlabs** [comment]
@resul4e yeah that was me, but i didn't implement tests for all methods in there. Some are kind of hard/impossible to test, but there are still some you can test. It probably wouldn't hurt to test methods which are not exposed to GDscript as well.

**Comment by resul4e** [comment]
@mbrlabs Ah okay good to know. However, as this is my first unit test, I'd like to start with something simpler.

**Comment by resul4e** [comment]
If it is all right, I would like to take a crack at testing the Bitmap class.

**Comment by Calinou** [comment]
> If it is all right, I would like to take a crack at testing the Bitmap class.

Sure, go ahead :slightly_smiling_face:

**Comment by DanielKriz** [comment]
Hello, I would like to make tests for Path3D. I also have two questions/suggestions:

1.  If I am using the TEST_SUITE feature of Doctest, Should I encapsulate all of Path3D unit test in one suite, or group my unit tests to different funtionality groups? I think it would be useful if we could settle on some format and stick to it.

2. I think it would be great if we organize tests in different folders (for example corresponding to scene/ directory, so path3d would be included as "scene/3d/test_path3d.h"), that way it would be much more readable and organized, even in test_main.cpp.

**Comment by Calinou** [comment]
> If I am using the TEST_SUITE feature of Doctest, Should I encapsulate all of Path3D unit test in one suite, or group my unit tests to different funtionality groups? I think it would be useful if we could settle on some format and stick to it.

We generally use only one test suite per class. We do use several test cases per class though.

**Comment by Xrayez** [comment]
> I think the current organization was chosen as to simplify the build system around it. I don't think having all tests in the same folder is much of a problem in practice.

To be honest, test cases with doctest can be written next to production code (as it was designed to be), but now we're talking about including test code in core, and I'm not sure whether this approach @reduz would be fond of.

But currently, it's possible to improve the build system to recursively collect test suite headers in `tests/`. This should also eliminate the need to include test suites in `test_main.cpp` manually.

**Comment by Gorgonx7** [comment]
If you don't mind I'm gonna take a look at testing `http_client`, I'm also going to refactor it a little to make it more testable no major changes just breaking down those larger functions a little

**Comment by jcsho** [comment]
Hi, interested in tackling unit tests for `shortcut`. 

Just to clarify it's the component in `scene/gui/shortcut.h`?

Let me know if there's anything I should be aware of or missed in the long thread!

**Comment by Calinou** [comment]
> Hi, interested in tackling unit tests for `shortcut`.
> 
> Just to clarify it's the component in `scene/gui/shortcut.h`?

It is :slightly_smiling_face:

**Comment by Hotspotmar** [comment]
I'm going to take a shot at doing the unit test for the `ArrayMesh`.

**Comment by im-deepfriedwater** [comment]
want to try to take on `SpriteFrames` if that's alright!

**Comment by Calinou** [comment]
> want to try to take on `SpriteFrames` if that's alright!

Sure, go ahead :slightly_smiling_face:

**Comment by im-deepfriedwater** [comment]
PR up for the SpriteFrames unit tests! 🙂 #57742

**Comment by skimmedsquare** [comment]
@Calinou - Sorry, work on my end stalled on `ImageTexture`, found myself going down a rathole of a bunch of different dependencies in making effective unit tests for that class due to its dependency on other classes / lack of mocking support within the project when I was looking. Feel free to remove my assignment from it for now.

**Comment by UltraTimon** [comment]
Hi All, I'd like to write tests for `Shortcut`. @resul4e mentioned [here](https://github.com/godotengine/godot/issues/43440#issuecomment-953595023) that he  worked on that, but when I checked out his work, it turns out he actually worked at `Bitmap`. So, for anyone wanting to work on that: [here](https://github.com/resul4e/godot/commit/81461c5c8fc83dd8b4557b3d9ff30709b22263ad) is part of that work. You can continue working on that if you want.

That being said, I will work on writing test cases for the `Shortcut` class. You can find my work [here](https://github.com/UltraTimon/godot/tree/shortcut-tests), so if I somehow don't finish this and don't respond anymore, you can continue where I left off.

I'll ask questions in the rocketchat if I have them!

**Comment by UltraTimon** [comment]
PR for the unit tests for Shortcuts is ready for review!
https://github.com/godotengine/godot/pull/58712

**Comment by kneejuicer** [comment]
If it's alright, I'd like to take a stab at writing some tests for `Geometry2D`.

**Comment by Calinou** [comment]
> If it's alright, I'd like to take a stab at writing some tests for the methods in `Geometry2D` that are still untested.

Sure, go ahead :slightly_smiling_face:

**Comment by UltraTimon** [comment]
Hi, I'd like to write tests for `InputEventKey`.

**Comment by Calinou** [comment]
> Hi, I'd like to write tests for `InputEventKey`.

Sounds good!

**Comment by UltraTimon** [comment]
The PR for `InputEventKey` is [ready for review](https://github.com/godotengine/godot/pull/59193)!

**Comment by StrawbDev** [comment]
Hi, I'd like to take on writing tests for AudioStreamSample.
@Calinou is the `save_to_wav()` method there as an example, or is it the only thing that needs testing?

**Comment by Calinou** [comment]
> @Calinou is the save_to_wav() method there as an example, or is it the only thing that needs testing?

I haven't checked for sure, but it might be the only thing that's easily unit testable right now.

**Comment by StrawbDev** [comment]
@Calinou The `set_data()` method on `AudioStreamSample` takes a lock on the `AudioServer` singleton, which is a problem as it isn't instantiated when the unit tests run. I can get it to work if I instantiate `AudioServer` along with `AudioDriverDummy`. Do you think it would be appropriate to add that to `test_main.cpp`? There are other servers being set up there in `test_case_start()`.

**Comment by StrawbDev** [comment]
[PR for AudioStreamSample tests](https://github.com/godotengine/godot/pull/60736) is now up for review!

**Comment by Calinou** [comment]
> I'm going to take a shot at doing the unit test for the `ArrayMesh`

@PrinceDeveloperOf Did you manage to get tests for ArrayMesh working in the end?

**Comment by PrinceDeveloperOf** [comment]
> > I'm going to take a shot at doing the unit test for the `ArrayMesh`
> 
> @PrinceDeveloperOf Did you manage to get tests for ArrayMesh working in the end? If not, no worries – let us know so we can update the task list. This way, another contributor can give it a look.

Let someone else do it.

**Comment by georgjz** [comment]
I'd like to give `Curve2D` a shot.

**Comment by PrinceDeveloperOf** [comment]
I'm trying to write the unit tests for `ArrayMesh`. When i try to call the method `add_surface_from_arrays()` It doesn't work because the `RenderingServer` singleton is null. There are ways to hack around from what I've been told, however it seems weird to hack out a unit test. Should i just work with what i can or maybe it moves too it's own issue?

**Comment by jamesfesik12** [comment]
Hey everyone! I am a first time open source contributor and wanted to take a shot at the CubeMap unit test! Where would I find the documentation and class setup for that unit test? Any help would be greatly appreciated.

**Comment by Calinou** [comment]
> Hey everyone! I am a first time open source contributor and wanted to take a shot at the CubeMap unit test! Where would I find the documentation and class setup for that unit test? Any help would be greatly appreciated.

CubeMap is a rarely used class, so there is no documentation for it. You'll have to read the source code and look at the methods defined in the class.

**Comment by jamesfesik12** [comment]
> 
I see thank you for your quick response! I have been researching inside the source code and have found the correct files to include however whenever I try to set the sides to a image color my IDE does not recognize the set_side method. Am I missing a file that I should include?

**Comment by jbcolli2** [comment]
I'd like to start working on unit tests for the Primitives, if no one else has started them.

**Comment by Calinou** [comment]
> I'd like to start working on unit tests for the Primitives, if no one else has started them.

Feel free to open a pull request for this :slightly_smiling_face: 

I'm not sure exactly *how* these should be tested though, so you may have to be creative here.

**Comment by jbcolli2** [comment]
Yeah, I'm new to the godot codebase, but I can't seem to find a place in the code where these classes are instantiated.  Am I missing something or are these classes just not currently used in the code?

**Comment by jbcolli2** [comment]
I've looked through the PRs for the tests that are checked off as done and I found a few that did not `#include` the test files in the `test_main.cpp` file.

- Math
- PathFollow2D
- PathFollow3D

Looks like the Math is maybe being worked on in the #40659, but the PathFollow are not mentioned in any other issue I can see.  Also when I do include `test_path_follow_2d.h` and `test_path_follow_3d.h` in `test_main.cpp` the build does not compile.  

Might be worth unchecking these unit tests and letting someone take a look at them again, unless this is already mentioned in another issue that I missed.  I'm currently working on the Primitive classes, but if the OP on this issue agrees, I'd be happy to take a look at the unit tests for PathFollow2D/3D after I'm done.

**Comment by raulsntos** [comment]
@jbcolli2 The PathFollow tests are being worked on https://github.com/godotengine/godot/pull/51372

**Comment by jbcolli2** [comment]
If no one else is working on it, I'll take a shot at tests for `ArrayMesh`.

**Comment by Geometror** [comment]
For those interested, here is a test coverage report generated with gcov/gcovr.
The thirdparty and tests directories are excluded as well as the generated header files (*.gen.h)
In case you want to generate it yourself and fiddle with it here are the steps:
1. Add --coverage (both as a LINKFLAG and CCFLAG) to the scons env and make sure you call scons with dev_build=yes and dev_mode=yes (or just tests=yes)
2. Run the godot executable with the --test argument andt wait until the tests are done
3. Generate the report with gcovr -e thirdparty/ -e tests/ -e .*\.gen\.h --html-details coverage.html

Of course there are still some files which don't make sense to include but overall I these numbers should be pretty accurate.

Compiling godot with --coverage (both as a LINKFLAG and CCFLAG) should be a scons build flag IMO so that we could generate coverage reports more conveniently (maybe even in CI).

Summary:
![image](https://user-images.githubusercontent.com/50084500/200307080-e5c5e940-acd5-4f28-bc19-e6011863454d.png)

Whole report (Google drive since GitHub does not allow uploading files over 25MB):
https://drive.google.com/file/d/149VlBwMaU_O5xChfwpdcOYJ8h-q6Jsgw/view?usp=sharing

**Comment by LucasLaukka** [comment]
@georgjz are you still working on Curve2D? Otherwise I was thinking about giving it a shot.

**Comment by georgjz** [comment]
@LucasLaukka I got side-tracked real badly 😄 have a go if you like, maybe I'll just move on to Curve3D....some day

**Comment by jainl28patel** [comment]
Hi @Calinou  I am new to Godot having experience in C++ dev.  Would like to work on `VisualShader` is no one is working on it. Can you please give some pointers on where to start from and what all to check ??

**Comment by Maran23** [comment]
> Hi @Calinou I am new to Godot having experience in C++ dev. Would like to work on `VisualShader` is no one is working on it. Can you please give some pointers on where to start from and what all to check ??

See the description.
Compiling instructions: https://docs.godotengine.org/en/latest/development/compiling/
Unit testing: https://docs.godotengine.org/en/latest/development/cpp/unit_testing.html
And then have a look on the other tests.

**Comment by purviporwal0** [comment]
Is this issue solved yet? Otherwise I'd like to work on it. Can you assign this issue to me?

**Comment by Maran23** [comment]
> Is this issue solved yet? Otherwise I'd like to work on it. Can you assign this issue to me?

This is only tracker of components which do not have any tests. It is a bit outdated, so you should checkout which components don't have any tests yet.

**Comment by ARez2** [comment]
Hi, first time contributor here,
I would like to get started with contributing to Godot by writing some tests. Since that one does not seem to be worked on yet, i am starting by trying to write some tests for `InputEventMouse`. Let's see how this goes!

**Comment by ARez2** [comment]
@Calinou i've noticed that both Curve2D and VisualShader Unit Tests have been Merged. Could you update the checklist at the top?

**Comment by Sauermann** [comment]
I have started working on Unit tests for `Viewport::_gui_input_event`.

**Comment by titus125** [comment]
Hi, new contributer here. I want to add some more methods in Transform3D covering methods that are currently not being tested such as the inverse functions and interpolate_with. Or is Transform3D no longer accepting new tests?

**Comment by Calinou** [comment]
> Hi, new contributer here. I want to add some more methods in Transform3D covering methods that are currently not being tested such as the inverse functions and interpolate_with. Or is Transform3D no longer accepting new tests?

Feel free to open a pull request to add more tests to an existing class :slightly_smiling_face:

**Comment by samkueb** [comment]
Hello! Another new contributor here. Was hoping to add some tests that raise the coverage of the Path2D and Projection classes.

**Comment by Calinou** [comment]
> Hello! Another new contributor here. Was hoping to add some tests that raise the coverage of the Path2D and Projection classes.

Feel free to open a pull request for these :slight

**Comment by SiddharthChillale** [comment]
Hello! This is my first time contributing! I'd like to add more coverage for Plane-Point operations for the Plane class specifically to test `distance_to()` method of the Plane class.

**Comment by SiddharthChillale** [comment]
I've started work on improving coverage of the `distance_of()` method of `Plane` class in #75602

**Comment by kgrubbs22** [comment]
Hello! I'm a first-time contributor and I would like to add unit tests for Curve3D and make sure the overload operators in Variant are all tested.

**Comment by SiddharthChillale** [comment]
I want to take up writing unit-tests for the `Cubemap` class. I've got the basic test file setup and will work on adding a basic test case to it. It should also serve as a good starting point for future test-cases for the `Cubemap` class. Please feel free to give suggestions or pitfalls to look for :)

**Comment by Calinou** [comment]
Testing CubeMap isn't feasible without RenderingServer access (which isn't available in the testing environment as headless builds don't instance a RenderingServer), so I'll strike it out from the list.

**Comment by SiddharthChillale** [comment]
I'm taking up the task of adding coverage for `HTTPClient` class.

**Comment by brno32** [comment]
`Path2D` is unchecked but wasn't it addressed in https://github.com/godotengine/godot/pull/66927?

**Comment by Calinou** [comment]
> `Path2D` is unchecked but wasn't it addressed in #66927?

Good catch, fixed :slightly_smiling_face:

**Comment by StefanoPetrilli** [comment]
Hey @samkueb, are you still working on Projection's tests? I am also starting to work on it. If you wish we can pair and tackle it together :)

**Comment by samkueb** [comment]
@StefanoPetrilli Nope, I ended up just working on Path2d. Feel free to work on Projection!

**Comment by brno32** [comment]
`DynamicFont` has been renamed `FontFile` in Godot 4, right? Should the checklist above reflect that?

**Comment by emgast** [comment]
I don't see anyone working on coverage for `AnimatedTexture` so I'm going to give it a shot.

**Comment by brno32** [comment]
`Input` itself doesn't have any tests. Should it be on here?

**Comment by azuloo** [comment]
Hi, I would like to see if we can cover `TranslationServer `. Is anybody working on it?

**Comment by noelg-cj** [comment]
Hey! First-time contributor here. I would like to work on ImageTexture. Please do tell me if someone is already working on it.

**Comment by noelg-cj** [comment]
Hey! I would like to work on `NoiseTexture`. Please do tell me if someone is already working on it. I'm new to unit testing and contributing and would really appreciate it if someone gave me pointers on where to start working on.

**Comment by Sauermann** [comment]
When using gcovr for coverage-reports as explained in https://github.com/godotengine/godot/issues/43440#issuecomment-1305518762, I had to experiment, in order to get it working with Clang. The following worked for me on Linux:

```
scons use_llvm=yes target=editor dev_mode=yes dev_build=yes use_coverage=yes '-j X'
./bin/godot.linuxbsd.editor.dev.x86_64.llvm --test
gcovr --gcov-executable "llvm-cov gcov" --exclude-directories thirdparty/ --exclude-directories tests/ -e .*.gen.h --decisions --html-details coverage.html
```

![image](https://github.com/godotengine/godot/assets/6299227/0514e9a8-e557-4482-bf8a-3b444b0aa3cc)

Line coverage went up from 16% to 23% since Nov 2022 🤩

**Comment by BrunoArmondBraga** [comment]
Hi everyone, I'm interested in making my first contribution with tests for the ReGex class. Just wanted to comment to make sure it's not being worked on!

**Comment by matorin57** [comment]
Howdy everyone. Looking to making my first contribution, just noting here that I'll be grabbing Font tests

**Comment by menteValidus** [comment]
I looked into AnimatedTexture class and for stable(4.1) it shows [Deprecated](https://github.com/godotengine/godot/blob/089ecc0a823cb61ef9b05a01a5ea1f01aea14689/doc/classes/AnimatedTexture.xml#L12). As it won't be updated and can be removed in future, is there still any need in tests for this class? Sorry, don't know who to tag here.

**Comment by Calinou** [comment]
> I looked into AnimatedTexture class and for stable(4.1) it shows [Deprecated](https://github.com/godotengine/godot/blob/089ecc0a823cb61ef9b05a01a5ea1f01aea14689/doc/classes/AnimatedTexture.xml#L12). As it won't be updated and can be removed in future, is there still any need in tests for this class? Sorry, don't know who to tag here.

I think we can still add tests for it. It has to remain supported throughout the 4.x cycle after all.

**Comment by menteValidus** [comment]
> > I looked into AnimatedTexture class and for stable(4.1) it shows [Deprecated](https://github.com/godotengine/godot/blob/089ecc0a823cb61ef9b05a01a5ea1f01aea14689/doc/classes/AnimatedTexture.xml#L12). As it won't be updated and can be removed in future, is there still any need in tests for this class? Sorry, don't know who to tag here.
> 
> I think we can still add tests for it. It has to remain supported throughout the 4.x cycle after all.

Ok, thanks. I'll work on this then.

**Comment by azuloo** [comment]
@Calinou Hi, can you please check out **Translation Server** ? It is covered by #79331. Thank you!

**Comment by menteValidus** [comment]
> I wrote some tests for ImageTexture which should be linked to this issue now (though they're not merged at the time of me writing this). I ran into a snag I was able to solve and wanted to call it out here.
> 
> Some of these classes require the `RenderingServer` for testing. At one point, that was not available and some of these tests were impossible to write, but I don't believe that is the case any longer - adding `[SceneTree]` to your unit test name would initialize the `RenderingEngine`:
> 
> https://github.com/godotengine/godot/blob/fe5b1c8d49313d63fbe91cb7cdf463e10fb86afa/tests/test_main.cpp#L223
> 
> I think that should be refactored to be more extensible (i.e. it doesn't make sense for random tests to add `[SceneTree]` to their name, but it does work.
> 
> Practically for this issue, I think that means `CubeMap` can be reopened as a thing that can be tested. I'm happy to prove that out myself.

I've encountered the same problem with `RenderingServer`. Actually, it's configured in the `test_main.cpp`. [Here's](https://github.com/menteValidus/godot/blob/29a118d7e6e9669facca2393698bd9e09e655d48/tests/test_main.cpp#L259) what I did in my fork to configure `RenderingServer` for my tests. And [here's](https://github.com/godotengine/godot/blob/c12d63556b5c1da03a00dd4c45c40e60bd8d68c2/tests/test_main.cpp#L237) how it looks in the original repository. You basically need to add `if (name.find("[YourTestingClass]") != -1) {` with configuration you need in the `test_main.cpp`.

**Comment by menteValidus** [comment]
I've encountered the same problem with RenderingServer. As you say it's configured in the test_main.cpp with `if (name.find("[SceneTree]") != -1) { `. I personally think that such configuration should happen in the test fixture itself but if we stick to the approach that is currently is in Godot tests - we need to add test environment configuration `if (name.find("[YourTestingClass]") != -1) { in the test_main.cpp.

[Here's](https://github.com/menteValidus/godot/blob/29a118d7e6e9669facca2393698bd9e09e655d48/tests/test_main.cpp#L259) what I did in my fork to configure RenderingServer for my tests.

**Comment by menteValidus** [comment]
@bgaudiosi, sorry forgot to mention you in my response

**Comment by AThousandShips** [comment]
AABB would benefit from more comprehensive testing of intersections, no testing is done for the intersection points for example

**Comment by Phoenixcresset** [comment]
Hey everyone, we are a small student group looking to make our first contribution.
We're looking to create a Unit Test for Camera2D, what is meant by "headless mode" ?

**Comment by AThousandShips** [comment]
Means there's no renderer active

**Comment by chocola-mint** [comment]
Hello, I'm currently working on tests for Camera3D. I'll be testing the methods that can work in headless mode, namely:

- `is_position_behind`
- `is_position_in_frustum`
- `project_local_ray_normal`
- `project_position`
- `project_ray_normal`
- `project_ray_origin`
- `unproject_position`

**Comment by SilicDev** [comment]
I'd be interested to work more on Variant tests. From what I can see the `NOT` operator has been tested already and there seem to be some `EQUAL` comparisions already via` identity_compare`, although the actual `operator` seems to use `hash_compare` (at least in C++). So I have a few questions:

1. Would it make sense to make another test case for `hash_compare`?
2. Should the comparision test cases only consider same type comparisions or should comparisions between i. e. Vector2 and Vector2i also be considered?
3. A lot of operators only apply to a subset of valid Variant types and often only for the same type, should a failure in other cases also be tested?

**Comment by chocola-mint** [comment]
I'm almost done with unit tests for Camera3D. There are just a few issues left to go over before I can submit the PR for review:

- As mentioned in #84007, there's a minor accuracy issue concerning one corner case with `unproject_position`. It's probably not a good idea to submit failing unit tests, so I'd like to confirm if it's intended behavior and should be ignored, or if I should wait for the issue to be fixed.
- Another problem is with how the "frustum" projection mode is very poorly documented and has several long-standing issues. (see: issue #61174, #75973, #63863 and #77906) It's not possible to design unit tests for this case at the moment, and even if I somehow succeeded, they would be failing and blocked by several issues anyway, so I've opted to skip this case and leave it for future contributors.

**Comment by BrickPi** [comment]
Planning on making a first contribution to add tests for `OptimizedTranslation`

**Comment by Calinou** [comment]
> From what I can tell, there is no `DynamicFont` class in the codebase and it was deleted in [99666de#diff-35b4cc3280195a381c5e95d48a19f9d95c968ffae2604e3cc6897d3261ca9afa](https://github.com/godotengine/godot/commit/99666de00fb30cb86473257776504ca70b4469c3#diff-35b4cc3280195a381c5e95d48a19f9d95c968ffae2604e3cc6897d3261ca9afa)
> 
> Should this be removed from the list above, or replaced with a different class?

`FontFile` is the 4.x replacement of `DynamicFont`. I've updated the list accordingly.

**Comment by wilsonchen02** [comment]
Hey everyone, I'm interested in trying to make some test cases for the AnimatedTexture class for my first contribution to Godot. I've seen some of the docs about unit tests and would like to know if there are any pointers to keep in mind for that class specifically. I saw in the ImageTexture PR #82086 that several test images were generated for the test suite, and I'm wondering if this needs to be done for AnimatedTexture as well. Thanks!

**Comment by menteValidus** [comment]
> Hey everyone, I'm interested in trying to make some test cases for the AnimatedTexture class for my first contribution to Godot. I've seen some of the docs about unit tests and would like to know if there are any pointers to keep in mind for that class specifically. I saw in the ImageTexture PR #82086 that several test images were generated for the test suite, and I'm wondering if this needs to be done for AnimatedTexture as well. Thanks!

Hey, I've actually wrote tests in my [fork](https://github.com/menteValidus/godot/blob/test-animated-texture/tests/scene/test_animated_texture.h) for the all fields of AnimatedTexture but got stuck on testing `_update_proxy` method. It depends on `OS::get_singleton()->get_ticks_usec()` and I wasn't sure at the moment how to mock it. Then I got busy at job and forgot about this for a month :) If you have an idea how to mock `OS` and write tests for `_update_proxy` I'll pass it to you. In `_update_proxy` happens all main logic and without it these tests I wrote are really "low value". On this week I'll attempt one more on cracking this one or asking help on dev chat.

**Comment by wilsonchen02** [comment]
Hey @menteValidus, apologies for not seeing your earlier comments in the issue😅. I am not familiar with mocking OS either. I'll avoid stepping on your toes and try taking a crack at GradientTexture instead. Thanks for the quick reply and letting me know about your progress.

**Comment by matorin57** [comment]
Do we currently have a runtime mock system? I’ve made some before where you can register mocks at test time via a string name (or you could use a function pointer interpreted as a int) and then add a line or two of macros at the top that in test builds that keeps track of keys to mock and if set will run some closure instead of the function itself

**Comment by alex-buchanan** [comment]
Hello All, as a new contributor, I'd like to tackle NoiseTexture, PackedScene, and PacketPeer.  If they're still available to claim, and what's the status on Font and what's still required to check off that test?

**Comment by Calinou** [comment]
> Hello All, as a new contributor, I'd like to tackle NoiseTexture, PackedScene, and PacketPeer.

Feel free to open pull requests for these :slightly_smiling_face: 

> and what's the status on Font and what's still required to check off that test?

Font has already been tackled by https://github.com/godotengine/godot/pull/81503. I've updated the list accordingly.

**Comment by jrobb-scottlogic** [comment]
Hello, I was looking to make my first contribution to the project and thought this would be a good issue to start with. Is there any particular unit test that would be best suited to a first time contributor or should I just dive in? Thanks :)

**Comment by jasperhale342** [comment]
> I'd be interested to work more on Variant tests. From what I can see the `NOT` operator has been tested already and there seem to be some `EQUAL` comparisions already via `identity_compare`, although the actual `==` operator seems to use `hash_compare` (at least in C++). So I have a few questions:
> 
> 1. Would it make sense to make another test case for `hash_compare`?
> 2. Should the comparision test cases only consider same type comparisions or should comparisions between i. e. Vector2 and Vector2i also be considered?
> 3. A lot of operators only apply to a subset of valid Variant types and often only for the same type, should a failure in other cases also be tested?

@SilicDev Did you start this one? I was going to try it if it's still open.

**Comment by TheSofox** [comment]
I linked this issue but just to state for the record: I added some unit tests for `UndoRedo` https://github.com/godotengine/godot/pull/85434

**Comment by ramadm** [comment]
@Phoenixcresset are you still working on Camera2D? I'd like to give it a try if not.

**Comment by Phoenixcresset** [comment]
> @Phoenixcresset are you still working on Camera2D? I'd like to give it a try if not.

We unfortunately reached a halt in our progress, so feel free to try your luck !
I can send you what we had done so far if you wish !

**Comment by ramadm** [comment]
> > @Phoenixcresset are you still working on Camera2D? I'd like to give it a try if not.
> 
> We unfortunately reached a halt in our progress, so feel free to try your luck ! I can send you what we had done so far if you wish !

That would be great, thank you.

**Comment by andrewargyrakis** [comment]
Hello, as a new contributor, I would like to try adding some unit tests for FontFile if that is still available.

**Comment by JustSascha** [comment]
Hi, a classmate and I would like to work on making tests for OptimizedTranslation for a class project an our first contribution. I don't see it mentioned in this thread but I might as well ask just in case: is anyone already taking care of this?

**Comment by atlas-wheatear** [comment]
Good day! I've written some (hopefully _sound_ 😅) unit tests for `CryptoKey`. Should I now make an MR to the `master` branch?

I saw a comment about not treading on toes, and may have taken it rather too literally.

[It's on this branch](https://github.com/atlas-wheatear/godot/tree/test-crypto-key). 😌

**Comment by wheatear-dev** [comment]
Hi @Calinou 👋 

A few things:

1. `CryptoKey` was addressed in #89021 . I can't update the issue, would you please be able to?
2. I've got a load of free time this weekend, and wanted to look at another unit-testing contribution.
3. `OptimizedTranslation` looks like a good candidate for my skillset, but I am unsure where @JustSascha got to with it.

**Comment by wheatear-dev** [comment]
Hi all 😃 

I'm working on a few tests for `JSONRPC` now.

**Comment by Calinou** [comment]
> 1. `CryptoKey` was addressed in https://github.com/godotengine/godot/pull/89021 . I can't update the issue, would you please be able to?

Done.

**Comment by fabricerenard12** [comment]
Hello, as a new contributor I would like to try adding some unit tests to Godot. Is there a class to test that would be a good fit for a new contributor?

**Comment by Magiccarpet00** [comment]
Hi, I'm @JustSascha 's classmate, and we're working on OptimizedTranslation. I don't understand why in the file test_translation.h there is this method:

`TEST_CASE("[OptimizedTranslation] Generate from Translation and read messages") {
	Ref<Translation> translation = memnew(Translation);
	translation->set_locale("fr");
	translation->add_message("Hello", "Bonjour");
	translation->add_message("Hello2", "Bonjour2");
	translation->add_message("Hello3", "Bonjour3");

	Ref<OptimizedTranslation> optimized_translation = memnew(OptimizedTranslation);
	optimized_translation->generate(translation);
	CHECK(optimized_translation->get_message("Hello") == "Bonjour");
	CHECK(optimized_translation->get_message("Hello2") == "Bonjour2");
	CHECK(optimized_translation->get_message("Hello3") == "Bonjour3");
	CHECK(optimized_translation->get_message("DoesNotExist") == "");

	List<StringName> messages;
	// `get_message_list()` can't return the list of messages stored in an OptimizedTranslation.
	optimized_translation->get_message_list(&messages);
	CHECK(optimized_translation->get_message_count() == 0);
	CHECK(messages.size() == 0);
}`

It only tests against French. But above all, I don't understand what else needs to be done, apart from adding other languages to the test?

**Comment by kosta777** [comment]
Hi, I'd like to add tests for FontFile, if nobody began working on it already. I see a message from @andrewargyrakis mentioning that they would do it, but it is from several months ago. Did you already start working on it or is it available?

**Comment by JustSascha** [comment]
Hi @wheatear-dev, sorry for not answering, I haven't been able to work much the pas few weeks so me and my classmate have not made too much progress on the unit test, but we're gonna work on that on the next few days.

If we get stuck we'll call for your help, thanks for offering!

**Comment by JonasSeerveld** [comment]
Hi, I'm a new contributor and currently a college student. I would like to take a shot at writing tests for the FontVariation class, if it is still available.

**Comment by BalmundSM** [comment]
Hello! I'm a new contributor and a college student like the guy above me. I'm gonna take AnimatedTexture, that is, if it's not already taken.

**Comment by gavenerickson** [comment]
Hey! I'm a new contributor and also a college student. I would like to try and write unit tests for GradientTexture.

**Comment by Calinou** [comment]
> Hey! I'm a new contributor and also a college student. I would like to try and write unit tests for GradientTexture.

There is already an open pull request for GradientTexture unit tests: https://github.com/godotengine/godot/pull/90501

If you're interested in adding tests for other Image/Texture-derived clases, we don't appear to have any tests for ImageTexture3D yet.

**Comment by gavenerickson** [comment]
Hi @Calinou, I will take a look at ImageTexture3D and thanks for linking the base cases for me to look at.

**Comment by SommAid** [comment]
Hello! I am a new contributor and was wondering if any tests OptimizedTranslation have been written?

**Comment by Calinou** [comment]
> Hello! I am a new contributor and was wondering if any tests OptimizedTranslation have been written?

I don't think anyone is writing tests for that class currently, so feel free to take a look :slightly_smiling_face:

**Comment by SommAid** [comment]
> > Hello! I am a new contributor and was wondering if any tests OptimizedTranslation have been written?
> 
> I don't think anyone is writing tests for that class currently, so feel free to take a look 🙂

I started to look into the OptimizedTranslation class, and testing and it appears that there is already a test suite for (get_message())[https://github.com/godotengine/godot/blob/master/tests/core/string/test_translation.h#L133]. Looking at the other methods some of them are missing support:
-  "The use of context is not yet supported in OptimizedTranslation." [get_message](https://github.com/godotengine/godot/blob/master/core/string/optimized_translation.cpp#L214)
- "// The use of plurals translation is not yet supported in OptimizedTranslation" [get_plural_message](https://github.com/godotengine/godot/blob/master/core/string/optimized_translation.cpp#L304)

Question: 
I can write a test for get_translated_message_list() or is there something else I should direct my attention towards?

**Comment by Calinou** [comment]
>  I can write a test for get_translated_message_list() or is there something else I should direct my attention towards?

I think you can write a new test for this method in the existing `test_translation.h`.

**Comment by Calinou** [comment]
> Hello, I am interested to write test cases for the Timer node. From looking at the timer class, im guessing i need to write cases for the get/set functions?

Yes, I think this is mostly all we can test in the Timer class.

**Comment by 2nafish117** [comment]
Hello, is anyone working on the Node2D tests? (Test the various exposed helper methods such as look_at() and move_local_x())

**Comment by Calinou** [comment]
> Hello, is anyone working on the Node2D tests? (Test the various exposed helper methods such as look_at() and move_local_x())

Not to my knowledge, so feel free to tackle these :slightly_smiling_face:

**Comment by mshojatalab** [comment]
Hi, I'm gonna be adding some tests for StyleBoxFlat.

**Comment by Magiccarpet00** [comment]
Hi, is anyone working on the AnimationPlayer tests ?

**Comment by rcooper47** [comment]
I plan to look into StyleBoxLine (and StyleBoxEmpty) tests.  Would it be a good idea to leave a link on the checklist above to people's forks when they start working on tests?

**Comment by mshojatalab** [comment]
> I plan to look into StyleBoxLine (and StyleBoxEmpty) tests. Would it be a good idea to leave a link on the checklist above to people's forks when they start working on tests?

That's taken, see above. 😊

**Comment by Calinou** [comment]
@mshojatalab @rcooper47 I've added you to the list in OP.

> Hi, is anyone working on the AnimationPlayer tests ?

@Magiccarpet00 There doesn't appear to be anyone working on tests for AnimationPlayer currently, so go ahead :slightly_smiling_face:

**Comment by powergainer** [comment]
Hello, @Calinou , I would like to contribute some unit tests here if that helps, what would be a good thing to do as a first time contributor? Thanks

**Comment by wheatear-dev** [comment]
Just working on a test file for `OptionButton` 😌

**Comment by OneBogdan01** [comment]
Hi, I have started working on the unit tests for the Decal node.

**Comment by lewyche** [comment]
I'm going to write some unit tests for Sky.

**Comment by WyattArnsdorf** [comment]
Hi @Calinou, I am looking to make my first contribution for a class project, so if there's a unit test that you'd like to get worked on a bit I'd really appreciate it!

**Comment by zedzee37** [comment]
Hello, is anyone here working on the unit tests needed for `Node2D`?

**Comment by Calinou** [comment]
> Hi @Calinou, I am looking to make my first contribution for a class project, so if there's a unit test that you'd like to get worked on a bit I'd really appreciate it!

You can pick any unit test you want that is not currently being worked on by someone else. I suggest picking a class you are already familiar with when developing Godot projects (other classes may be harder to test if you have no prior knowledge).

**Comment by LeonStansfield** [comment]
Hi, has anyone worked on the tests for the height_map_shape_3D? I am willing to work on them as a good first project!

**Comment by 2nafish117** [comment]
> Hello, is anyone here working on the unit tests needed for `Node2D`?

yep, i have tests written for the Node2D #91654.

@Calinou or other contributors, could you please review it once more and merge if it looks good. Thanks.

**Comment by LeonStansfield** [comment]
I would like to spend some time doing parallax_2d if thats okay?

**Comment by AidanPohl** [comment]
I'd like to throw my hat in on `FontFile`

**Comment by pafuent** [comment]
Hi @Calinou. Seems that the PR description should be updated to reflect that [this](https://github.com/godotengine/godot/pull/91395) PR implemented test for `Timer`

**Comment by pafuent** [comment]
I'm adding tests for `StreamPeer`  and `StreamPeerBuffer` while I'm trying to fix a bug on `StreamPeer` here #95784

**Comment by JayTropper** [comment]
I am going to write a test for `Node3D`.

**Comment by tetrapod00** [comment]
I'll be adding some unit tests for Control.

**Comment by ghildim** [comment]
Hey Guys, new to Open-Source stuff. How do I get started and contribute?

**Comment by tetrapod00** [comment]
@ghildim this issue isn't the right place to ask. You should take a look at [the official docs](https://docs.godotengine.org/en/latest/contributing/how_to_contribute.html) and join the [contributors chat](https://chat.godotengine.org/home). There's a new contributors channel there where you can ask questions.

**Comment by ghildim** [comment]
@tetrapod00 Thanks! Read the documentation, and I'm going to reread the thread and try to help

**Comment by tetrapod00** [comment]
@ghildim You might find it easier to get started by adding some unit tests to a class that already has some tests, before writing all the tests for a class that has none. That way you'll have example code to work from and can learn the codebase,

**Comment by D0V4HKIIN** [comment]
I will be working on adding unit tests for the PhysicsMaterial class.

**Comment by Gurka2** [comment]
I will be working on adding unit tests for the ReflectionProbe class.

**Comment by pafuent** [comment]
@Calinou Can I get some reviews on Unit Tests PRs?
https://github.com/godotengine/godot/pull/95590
https://github.com/godotengine/godot/pull/95784
https://github.com/godotengine/godot/pull/95931
https://github.com/godotengine/godot/pull/97262
https://github.com/godotengine/godot/pull/97311

**Comment by dustdfg** [comment]
I have a question about `SceneState` which is declared along `PackedScene`. Doesn't it need tests too?

**Comment by girdenis-p** [comment]
I'm going to start writing some unit tests for Control to do with general layout e.g. anchoring, resizing etc.

**Comment by CampioniMan** [comment]
Hello, I am going to create some unit tests for `WebSocketPeer`!

**Comment by NeuralTide** [comment]
I want to work on my first contribution. I will start working on adding tests for Parallax2D.

**Comment by carsonetb** [comment]
I would like to do tests for OccluderInstance3D for my first contribution. Thanks!

**Comment by JustinBraben** [comment]
I wish to work on my first contribution. I can work on `FontFile`.

**Comment by qwertychomp** [comment]
I want to make tests for Occluder3D and all the occluder shapes (BoxOccluder3D, SphereOccuder3D, etc).

**Comment by JustinBraben** [comment]
Closed #99406 since #99131 was merged in. May I work on `FontVariation` instead?

**Comment by Flarkk** [comment]
Core class `Projection` now covered in #99125.
Open for reviews !

**Comment by ZhiyiHu93** [comment]
Hi, if it hasn't been taken by others, I want to add some tests to Sky.

**Comment by qwertychomp** [comment]
My Occluder shapes unit test should finally be ready to merge now, someone please take a look at it.

**Comment by Sauermann** [comment]
Line coverage of all tests went up from 23% to 27.8% since [Aug 2023](https://github.com/godotengine/godot/issues/43440#issuecomment-1667606134) 🤩 Great job everyone!

![Image](https://github.com/user-attachments/assets/bd251e75-ce9c-46cb-bd2b-51e621d0cc67)

**Comment by qwertychomp** [comment]
Also, can i please work on unit tests for GLTFState/GLTFDocument?

**Comment by carsonetb** [comment]
By the way, this has been ready to merge for a while, but a review is needed (: 
https://github.com/godotengine/godot/pull/99379

**Comment by JulianHeuser** [comment]
I've created a PR for GLTFDocument: #101136. @qwertychomp if you're still working on one yourself, feel free to continue - I'm a new contributor so if you have anything it's probably better.

**Comment by fire** [comment]
@Sauermann 

> Line coverage of all tests went up from 23% to 27.8% since https://github.com/godotengine/godot/issues/43440#issuecomment-1667606134 🤩 Great job everyone!

How do you calculate this?

**Comment by tsarquis88** [comment]
I've started working on UTs for `StreamPeerGzip`, hope to raise a PR soon ;)

**Comment by Sauermann** [comment]
@fire 
These numbers are calculated with [gcovr](https://gcovr.com).
For this to work, Godot needs to be built with the scons option `use_coverage=yes`, then all tests need to be run and afterwards gcover creates a set of html documents with all coverage infos.
See [this comment](https://github.com/godotengine/godot/issues/43440#issuecomment-1667606134) for details about this process on Linux.

**Comment by mareknovakz** [comment]
I can try write tests for httprequest.

**Comment by AThousandShips** [comment]
There is already a PR open for Http request:
* https://github.com/godotengine/godot/pull/99687

**Comment by pafuent** [comment]
FYI @mareknovakz 

> I can try write tests for httprequest.

I wrote tests for `HTTPRequest` here https://github.com/godotengine/godot/pull/95224
That PR tries two different approaches to test it, one using a mocking library and other that involves implementing a mock by hand

**Comment by RafaDdS** [comment]
I'm considering writing tests for VisibleOnScreenNotifier2D and 3D after a little research it looks like the problem would be emulating RendererCanvasCull::canvas_item_set_visibility_notifier and  RendererCanvasCull::update_visibility_notifiers while the engine is in headless. Is that right? Can't these methods just be mocked?

**Comment by wheatear-dev** [comment]
Working on unit tests for `StreamPeerTCP`. 😌

**Comment by fire** [comment]
Since we have the technology from gltfdocument tests to save and load resources. 

How would we test StandardMaterial3D (and ORMMaterial3D)? I assume it would need to be headless? We could check existence of properties after save and load. Does’t seems satisfying since the result of a material is a shader stack resulting in a camera look.

**Comment by Calinou** [comment]
> How would we test StandardMaterial3D (and ORMMaterial3D)? I assume it would need to be headless? We could check existence of properties after save and load. Does’t seems satisfying since the result of a material is a shader stack resulting in a camera look.

Yes, all we can do within an unit test is call various setters and getters (along with the conversion to ShaderMaterial).

Testing the actual material rendering is better done using a visual regression testing project like https://github.com/Calinou/godot-rendering-tests. I have a branch to integrate it with our CI: https://github.com/Calinou/godot/tree/ci-check-visual-regressions

**Comment by jmii1236** [comment]
hi, would i be able to work on control testing?

**Comment by michelleboisvert** [comment]
Hello, fellow newbie here!

I would be interested in working on unit tests for the AudioStreamInteractive class. Hopefully this is okay, but if not please let me know!

**Comment by yuantianle** [comment]
Hello, team!

I am interested in working on UTs for the StandardMaterial3D class. Asking for a seat here. Thanks!

**Comment by TheShed412** [comment]
I have a PR for the `StyleBoxFlat` component https://github.com/godotengine/godot/pull/106009

**Comment by qwertychomp** [comment]
My PR that includes unit tests for Occluder shapes, #108891 , is ready to merge! i made a new one after the first one from 8 months ago because i accidentally broke that one when fiddling around with rebase (which i forgot how to use :(  )

**Comment by adriano-sudario** [comment]
hey!
is there anyone avaiable to take a look at these opened PRs?

**Comment by Clarkson1415** [comment]
Hey, haven't contributed to Godot yet but would like to try write some tests, seems like the `StreamPeerTLS` is one of the few remaining on that list so I'll have a look at that one if no one else is currently.

**Comment by WatcherMagic** [comment]
Hi! I'd like to write tests for ENetConnection and ENetPacketPeer, though networking is not my forte and I'm sure I'll be learning a lot along the way so if there's something else available I can help with that instead.

**Comment by wheatear-dev** [comment]
Hi, me again 👋 

I have time on my hands this weekend, and want to add additional tests for `Control`.

I see a couple of PRs for this already:

- #98757
- #107037

It looks like a huge component! So, are there any gaps between these two? (cc @Calinou)

**Comment by fire** [comment]
I don't think FontVariation was complete?

**Comment by wheatear-dev** [comment]
> I don't think FontVariation was complete?

Thank you @fire , I'll work on that then.

**Comment by albertok** [comment]
ENetConnection (and ENetPacketPeer) are [done](https://github.com/godotengine/godot/pull/112984). Just needs merging.

@Calinou Happy to do StreamPeerTLS and we can close this big ticket off. :)

**Comment by Clarkson1415** [comment]
A while ago said I was working on StreamPeerTLS, didn't get to it, got busy with other things. Apologies. Feel free to : )

**Comment by Calinou** [comment]
> Hey, first time contributor and new to open source in general, I'd like to take a look at writing some unit tests for OccluderInstance3D if it's still open!

OccluderInstance3D has tests implemented in https://github.com/godotengine/godot/pull/116234 now. It's no longer up for grabs, but you can check the first post for classes that are still up for grabs (I recently updated the list).

**Comment by ioanntza22** [comment]
Hi! I've opened a PR adding unit tests for `AStarGrid2D`, which wasn't on the tracker list but had zero test coverage (the sibling `AStar3D` class had tests but `AStarGrid2D` did not):

- **PR:** #117884
- **24 test cases / 96 assertions** covering: initialization defaults, region/dirty flag behavior, bounds checking (including negative offsets), solid points, weight scales, position calculation, and pathfinding (obstacles, blocked paths, partial paths, diagonal modes, weight-based detours).

Happy to address any feedback from reviewers!

**Comment by Spudoku** [comment]
Currently planning on adding tests for `Control (container logic)`, `ENetConnection` and `ENetPacketPeer`, and/or `StreamPeerTLS`. Just want to confirm that these are still available for contribution.

**Comment by Calinou** [comment]
There are already open PRs for StreamPeerTLS, as well as ENetConnection + ENetPacketPeer:
 
- https://github.com/godotengine/godot/pull/102961
- https://github.com/godotengine/godot/pull/111784
- https://github.com/godotengine/godot/pull/112984

There is also a PR for Control container logic:

- https://github.com/godotengine/godot/pull/117278

I've updated OP accordingly. Since all items are checked now, I'll go through again the list of all engine classes next week to find more test-friendly classes that can be added to the list.

**Comment by Spudoku** [comment]
Ok, thanks for getting back to me @Calinou

**Comment by Spudoku** [comment]
Is there an established/efficient way for contributors to find classes to write tests for?

**Comment by Calinou** [comment]
> Is there an established/efficient way for contributors to find classes to write tests for?

This can't be automated, it needs manual decision based on a few factors:

- Whether the class depends on rendering or physics.
  - Neither rendering nor physics are deterministic, and are therefore a bad fit for unit testing.

- Whether the class needs DisplayServer or RenderingServer to function.
   - DisplayServer and RenderingServer are not available in headless mode, which makes them untestable with unit tests (it always uses headless mode). Editor-only classes are also not currently unit testable.

- Whether the class is friendly to automated testing.
  - Good candidates are self-contained classes with no dependencies on other classes.
  - Even better candidates are classes that expose methods that can take arbitrary user input, which may be invalid or malformed.

**Comment by Spudoku** [comment]
I found that some methods of some classes already marked as "completed" are missing tests, for example:

1. `Geometry 2d` is missing tests for `triangulate_delaunay`, `triangulate_polygon` and `offset_polygon`
2. `Image` is missing tests for `rotate_90`, `rotate_180`, `normal_map_to_xy`, `linear_to_srgb`, `decompress`, and `detect_alpha`
3. `node2d` is missing tests for `rotate`, `translate` and `apply_scale`

It is entirely possible that there are good reasons that the aforementioned methods are not testable. If so, please let me know. Otherwise, please advise on which methods may benefit from unit tests.

As for classes completely lacking tests:
1. `script_debugger`
2. `ustring`

Again, there may be good reasons that these lack tests.

The way I found these was looking through the files of my fork of the `master` branch, reading through official documentation and combing the test files.

**Comment by Spudoku** [comment]
@Calinou 
Hello again. Is there an update on the classes and/or methods I've listed above?

**Comment by calbebop** [comment]
Not @Calinou, but figured I'd chime in as a new contributor to [#43440](https://github.com/godotengine/godot/issues/43440).

Short version on your gaps: the "Completed classes" preamble actually addresses this directly:
> Further improvements may be possible by testing methods that were added after the tests were merged.

So the missing `Image`, `Geometry2D`, and `Node2D` methods you found all read to me as fair game.

On the broader "how do I find candidates" question, here's the loop I've personally been running for the recent GUI tests I've contributed, of which, about half weren't on the intial tracker:

1. **Diff the directories:** `ls scene/gui/*.h` against `ls tests/scene/test_*.cpp`. Anything without a corresponding test file is a candidate. The tracker is a curated subset, not the full gap.
2. **Apply @Calinou's filter above**: rendering, physics, and `DisplayServer` dependencies rule out a lot quickly.
3. **Skim the implementation .cpp:** Count public methods that aren't `_draw()` or `_notification()` glue. Mode flags, structural mutators, and signal emit paths are where the real coverage lives. If the surface is mostly draw plumbing, skip.
4. **Grep open PRs:** Look for the _snake_case_ filename before claiming.
5. **The lens I try to come back to is "quality over fluff"**: Prefer classes where you can assert mode-specific contracts, multi-step invariants, or signal emit paths over getter/setter round-trip farms.

Practically, for your list I'd just open small standalone PRs, @Spudoku. One per class or tight group without waiting for go-ahead here. In my opinion, small well-scoped test PRs are easier to have discussions around.

**Comment by Spudoku** [comment]
Thanks for the clarification and assistance, @calbebop!

**Comment by Halei-6103** [comment]
I'd like to write unit tests for StyleBoxFlat as the existing PR #106009 appears to be stalled and unmerged

**Comment by rohithnafeel** [comment]
Hi! I'd like to work on unit tests for StreamPeerTLS, if it's still available. This will be my first contribution to Godot.

**Comment by AThousandShips** [comment]
> Hi! I'd like to work on unit tests for StreamPeerTLS, if it's still available. This will be my first contribution to Godot.

It does not seem to be available, two open PRs are listed for that in the list, check those and see

**Comment by Grace-Shao** [comment]
Hello! I'd like to work on unit tests for **FontVariation** if it's still available since the two PRs for it are closed. Also a first time contributor to Godot but love the engine.

**Comment by Calinou** [comment]
> Hello! I'd like to work on unit tests for **FontVariation** if it's still available since the two PRs for it are closed. Also a first time contributor to Godot but love the engine.

Feel free to open a PR for this 🙂

**Comment by jsudol85** [comment]
Hi. Just to double check, aside from FontVariation, it appears that at this time no PR's are required for anything under Classes to test, unless a new class is listed or an issue arises with a current PR? Please let me know. Thank you again!

**Comment by Calinou** [comment]
> Hi. Just to double check, aside from FontVariation, it appears that at this time no PR's are required for anything under Classes to test, unless a new class is listed or an issue arises with a current PR? Please let me know. Thank you again!

I haven't gotten time to update the list of testable classes lately, but there are definitely still classes that could be tested; they're just not listed here yet.

**Comment by krubka1** [comment]
Hello, I was looking at test units for test_node_2d.cpp. It has Test cases for look_at and similar but some of the setters and getters do not have TestCases or Subcases under other Testcases. so if no one is working on them, I would like to do so as I want to contribute to Godot.

**Comment by AThousandShips** [comment]
What setters and getters are you talking about? Setters and getters don't get tests in general only if they are relevant or needed

**Comment by krubka1** [comment]
set_skew()/get_skew(), set_rotation_degree() and so on, I was just new to it so I wanted to start simple.

**Comment by AThousandShips** [comment]
I'd suggest looking at cases where there's more specific things missing or classes lacking tests, deciding what methods need tests is a bit complicated and we should avoid adding them unless there's something specific to catch

**Comment by krubka1** [comment]
thanks, I will look for it.

**Comment by jsudol85** [comment]
Would a class like AbstractYuyvBufferDecoder be a good candidate for testing coverage?

**Comment by LeonxLJX** [comment]
I'd like to take this one.

Reading the request, this is a capability gap rather than a bug: `ui_focus_` doesn't currently cover the [TRACKER] Unit tests to add or improve case you described. The detail you gave — especially the `ui_focus_` aspect — is the constraint I'd design around.

Plan: map the existing implementation of the closest related option to understand the conventions, then add the new capability mirroring that shape — with types and tests if the repo has coverage for this area, plus a doc note if the feature is user-facing.

One thing I'd like to confirm before starting: for [TRACKER] Unit tests to add or improve, should the behavior be opt-in (default unchanged) or change the default? I'd rather keep backwards compatibility unless the repo prefers otherwise.

(claiming via @LeonxLJX)

**Comment by anr1996** [comment]
Hello, I am a new contributor. This will be my first time! I was interested in writing unit tests.  I thought it would be a good start. I'm somewhat new to C++. I was looking at StyleBoxFlat . However, I noticed that it is already being worked on. Would OptimizedTranslation be a good place to start writing unit tests?
